### PR TITLE
[WIP] Heron client unit test

### PR DESCRIPTION
**Pull request by objmagic**
We test if we will `Popen` two Java program which writes definition file and calls submitter. I can add more tests but I'd like to have some opinions first.

### CODE REVIEW

File: heron/tools/cli/src/python/execute.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def heron_class(class_name, lib_jars, extra_jars=None, args=None, java_defines=N
   all_args += [class_name] + list(args)
 
   # set heron_config environment variable
-  heron_env = os.environ.copy()
+  heron_env = {}
```
Comment (billonahill) [inline heron/tools/cli/src/python/execute.py]: this seems like a functional change beyond just what's required for the testing, no?

File: heron/tools/cli/src/python/args.py [hunk 2]
```diff
@@ -41,18 +41,6 @@ def add_verbose(parser):
   return parser
 
 
-def add_trace_execution(parser):
-  '''
-  :param parser:
-  :return:
-  '''
-  parser.add_argument(
-      '--trace-execution',
```
Comment (billonahill) [inline heron/tools/cli/src/python/args.py]: why is this being removed?

File: heron/tools/cli/tests/python/parse_unittest.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    result.SimpleResult.__init__ = Mock(return_value=None)
+    result.ProcessResult.__init__ = Mock(return_value=None)
+    def launch(cl_args, topology_file, tmp_dir):
+      return submit.launch_a_topology(cl_args, tmp_dir, topology_file, 'T.defn', 'WordCount')
+    submit.launch_topologies = Mock(side_effect=launch)
+    pass
+
+  def test(self):
+    subprocess.Popen = MagicMock()
+    args = ["heron", "submit", "local", "~/.heron/examples/heron-examples.jar",
```
Comment (billonahill) [inline heron/tools/cli/tests/python/parse_unittest.py]: If we're going to copy this pattern for other calls, we might want to consider a generic method that takes a string of the heron command and a collection of strings of popen commands. Using strings will likely be easier than arrays. The shared method would deal with converting strings to arrays.

File: heron/tools/cli/tests/python/parse_unittest.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import os
+import subprocess
+import sys
+import tempfile
+import heron.tools.cli.src.python.main as main
+import heron.tools.cli.src.python.submit as submit
+import heron.tools.cli.src.python.result as result
+import heron.tools.common.src.python.utils.config as config
+
+class OptsTest(unittest.TestCase):
```
Comment (billonahill) [inline heron/tools/cli/tests/python/parse_unittest.py]: `OptsTest` sounds like a test that verifies option parsing logic. Can we call this something more representative like `SubmitTest` or something (which extends a common `ClientCommandTest` or something like that)? We can then have a bunch of tests to represent how different args to submit result in different popens.

File: heron/tools/cli/tests/python/parse_unittest.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    config.parse_override_config = MagicMock(return_value='/heron/home/override.yaml')
+    # Mock result module
+    result.render = MagicMock(return_value=None)
+    result.is_successful = MagicMock(return_value=True)
+    result.SimpleResult.__init__ = Mock(return_value=None)
+    result.ProcessResult.__init__ = Mock(return_value=None)
+    def launch(cl_args, topology_file, tmp_dir):
+      return submit.launch_a_topology(cl_args, tmp_dir, topology_file, 'T.defn', 'WordCount')
+    submit.launch_topologies = Mock(side_effect=launch)
+    pass
```
Comment (billonahill) [inline heron/tools/cli/tests/python/parse_unittest.py]: is `pass` necessary?

File: heron/tools/cli/src/python/args.py (hunk 2 above)
Comment (objmagic) [inline heron/tools/cli/src/python/args.py]: It is not used. Pretty sure it is some leftover when we move from close-sourced to open-sourced Heron.

File: heron/tools/cli/src/python/execute.py (hunk 1 above)
Comment (objmagic) [inline heron/tools/cli/src/python/execute.py]: At first I thought we don't need to copy all the env variables, but now I think we do need all env vars. For example, Java program could fetch environmental variables.

Comment (billonahill) [inline heron/tools/cli/src/python/execute.py]: right, so should this change be reverted?

File: heron/tools/cli/tests/python/parse_unittest.py (hunk 3 above)
Comment (objmagic) [inline heron/tools/cli/tests/python/parse_unittest.py]: addressed in https://github.com/twitter/heron/pull/1983/commits/f18b0a417bb96f0b473db62eea39b12705a59f3f. Using string looked better at first. But when I was refactoring to use string, I realized the command string is too long and I have to break them into multiple lines, but I forgot to add **a space** at the end of some lines, which made me spend like 5 minutes comparing why assertion failed. I suggest we use **list**, which makes spacing clearer.
image

File: heron/tools/cli/src/python/execute.py (hunk 1 above)
Comment (objmagic) [inline heron/tools/cli/src/python/execute.py]: yes later.

Comment (objmagic) [inline heron/tools/cli/src/python/execute.py]: https://github.com/twitter/heron/pull/1983/commits/4de4efbb19ade9fada1a6adbe6c9b6bd7e8cbfd7

File: heron/tools/cli/tests/python/parse_unittest.py (hunk 3 above)
Comment (billonahill) [inline heron/tools/cli/tests/python/parse_unittest.py]: IMO putting spaces at the end of multi-line strings is a pretty standard thing to do, so I would recommend that.

We could even evolve this to read the commands from a text file at some point if we wanted to remove the need for long strings in code.

**Comment by objmagic** [comment]
@billonahill if this looks good now, we can add more tests now.

**Comment by billonahill** [comment]
This looks good. Can you merge this and then add more tests in another PR?

**Pull request merged**